File: frontend/src/services/adtService.ts
import type { Admission, Bed, Room, Ward } from '../types/adt';
import api from './api';

export const ADTService = {
  // Wards
  createWard: async (ward: Partial<Ward>): Promise<Ward> => {
    const response = await api.post<Ward>('/wards', ward);
    return response.data;
  },

  listWards: async (): Promise<Ward[]> => {
    const response = await api.get<Ward[]>('/wards');
    return response.data;
  },

  // Rooms
  createRoom: async (room: Partial<Room>): Promise<Room> => {
    const response = await api.post<Room>('/rooms', room);
    return response.data;
  },

  // Beds
  createBed: async (bed: Partial<Bed>): Promise<Bed> => {
    const response = await api.post<Bed>('/beds', bed);
    return response.data;
  },

  listBeds: async (status?: string): Promise<Bed[]> => {
    const params = status ? { status } : {};
    const response = await api.get<Bed[]>('/beds', { params });
    return response.data;
  },

  // Admissions
  admitPatient: async (admission: Partial<Admission>): Promise<Admission> => {
    const response = await api.post<Admission>('/admissions', admission);
    return response.data;
  },

  dischargePatient: async (id: number): Promise<void> => {
    await api.post(`/admissions/${id}/discharge`);
  },

  listActiveAdmissions: async (): Promise<Admission[]> => {
    const response = await api.get<Admission[]>('/admissions/active');
    return response.data;
  },

  getAdmission: async (id: number): Promise<Admission> => {
    const response = await api.get<Admission>(`/admissions/${id}`);
    return response.data;
  },

  // Transfers
  transferPatient: async (data: any): Promise<void> => {
    await api.post('/transfers', data);
  },

  // Discharge Summaries
  createDischargeSummary: async (data: any): Promise<void> => {
    await api.post('/discharge-summaries', data);
  },

  getDischargeSummary: async (admissionId: number): Promise<any> => {
    const response = await api.get(
      `/admissions/${admissionId}/discharge-summary`
    );
    return response.data;
  },
};

File: backend/internal/service/adt_service.go
package service

import (
	"errors"
	"time"

	"github.com/zarishsphere/zarish-his/internal/models"
	"github.com/zarishsphere/zarish-his/internal/repository"
)

type ADTService struct {
	repo *repository.ADTRepository
}

func NewADTService(repo *repository.ADTRepository) *ADTService {
	return &ADTService{repo: repo}
}

func (s *ADTService) CreateWard(ward *models.Ward) error {
	return s.repo.CreateWard(ward)
}

func (s *ADTService) ListWards() ([]models.Ward, error) {
	return s.repo.ListWards()
}

func (s *ADTService) CreateRoom(room *models.Room) error {
	return s.repo.CreateRoom(room)
}

func (s *ADTService) CreateBed(bed *models.Bed) error {
	return s.repo.CreateBed(bed)
}

func (s *ADTService) ListBeds(status string) ([]models.Bed, error) {
	return s.repo.ListBeds(status)
}

func (s *ADTService) AdmitPatient(admission *models.Admission) error {
	// Validate bed availability
	beds, err := s.repo.ListBeds("Available")
	if err != nil {
		return err
	}

	isAvailable := false
	for _, bed := range beds {
		if bed.ID == admission.BedID {
			isAvailable = true
			break
		}
	}

	if !isAvailable {
		return errors.New("selected bed is not available")
	}

	admission.AdmissionDate = time.Now()
	admission.Status = "Admitted"
	return s.repo.CreateAdmission(admission)
}

func (s *ADTService) DischargePatient(admissionID uint) error {
	return s.repo.DischargePatient(admissionID)
}

func (s *ADTService) ListActiveAdmissions() ([]models.Admission, error) {
	return s.repo.ListAdmissions("Admitted")
}

func (s *ADTService) GetAdmission(id uint) (*models.Admission, error) {
	return s.repo.GetAdmission(id)
}

// TransferPatient transfers a patient to a new ward/bed
func (s *ADTService) TransferPatient(transfer *models.Transfer) error {
	// Validate destination bed is available
	beds, err := s.repo.ListBeds("Available")
	if err != nil {
		return err
	}

	isAvailable := false
	for _, bed := range beds {
		if bed.ID == transfer.ToBedID {
			isAvailable = true
			break
		}
	}

	if !isAvailable {
		return errors.New("destination bed is not available")
	}

	transfer.TransferDate = time.Now()
	return s.repo.CreateTransfer(transfer)
}

func (s *ADTService) ListTransfers(admissionID uint) ([]models.Transfer, error) {
	return s.repo.ListTransfers(admissionID)
}

// CreateDischargeSummary creates a discharge summary and updates admission status
func (s *ADTService) CreateDischargeSummary(summary *models.DischargeSummary) error {
	summary.DischargeDate = time.Now()

	// Create the summary
	if err := s.repo.CreateDischargeSummary(summary); err != nil {
		return err
	}

	// Discharge the patient
	return s.repo.DischargePatient(summary.AdmissionID)
}

func (s *ADTService) GetDischargeSummary(admissionID uint) (*models.DischargeSummary, error) {
	return s.repo.GetDischargeSummary(admissionID)
}
